exec/java-exec/src/test/java/org/apache/drill/test/ClusterTest.java

```diff
@@ -125,4 +132,56 @@ public static void run(String query, Object... args) throws Exception {
   public QueryBuilder queryBuilder( ) {
     return client.queryBuilder();
   }
+
+  /**
+   * Utility method which tests given query produces a {@link UserException} and the exception message contains
+   * the given message.
+   * @param testSqlQuery Test query
+   * @param expectedErrorMsg Expected error message.
+   */
+  protected static void errorMsgTestHelper(String testSqlQuery, String expectedErrorMsg) throws Exception {
+    try {
+      run(testSqlQuery);
+      fail("Expected a UserException when running " + testSqlQuery);
+    } catch (UserException actualException) {
+      try {
+        assertThat("message of UserException when running " + testSqlQuery, actualException.getMessage(), containsString(expectedErrorMsg));
+      } catch (AssertionError e) {
+        e.addSuppressed(actualException);
+        throw e;
+      }
+    }
+  }
+
+  protected static void updateClient(Properties properties) {
+    if (client != null) {
+      client.close();
+      client = null;
+    }
+    ClientFixture.ClientBuilder clientBuilder = cluster.clientBuilder();
+    if (properties != null) {
+      for (final String key : properties.stringPropertyNames()) {
+        final String lowerCaseKey = key.toLowerCase();
+        clientBuilder.property(lowerCaseKey, properties.getProperty(key));
+      }
+    }
+    client = clientBuilder.build();
+  }
+
+  protected static void updateClient(final String user) {
+    updateClient(user, null);
+  }
+
+  protected static void updateClient(final String user, final String password) {
+    if (client != null) {
+      client.close();
+      client = null;
+    }
+    final Properties properties = new Properties();
+    properties.setProperty(DrillProperties.USER, user);
+    if (password != null) {
+      properties.setProperty(DrillProperties.PASSWORD, password);
+    }
+    updateClient(properties);
+  }
```

Comment:
Let's move these additions to ClientFixture. It will mean that they're no longer drop-in replacements but they'll be more at home there.

Reply:
HI James, ClientFixture doesn't have client variable, so that we could't close client at first. And cluster here is a non-static variable, we should change all updateClient static method to non-static method, should we still keep changing this?

Reply:
> Let's move these additions to ClientFixture. It will mean that they're no longer drop-in replacements but they'll be more at home there.

Hi James, ClientFixture doesn't have a client variable there, so that we could't close client at first and we need to change updateClient signature from void to return ClientFixture, besides due to cluster variable here is none static, we also need to change updateClient from static method to none static method. Furthermore, there are still lots of test cases that we could just update client without restart cluster, and for those cases that do need restart cluster, we could use startCluster method to config new cluster and client at same time without using updateClient. So maybe we could still keep those code here?

Reply:
Ah, yes you're right. These methods still look a bit out of place. Can errorMsgTestHelper move to ClientFixture? And can the new updateClient methods be replaced by usage of cluster.addClientFixture(Properties p) and cluster.client(int number)?

Reply:
Let me test out a couple of things out locally. If they work I'll send a PR to you, otherwise we'll be ready to merge as is.

Reply:
Thanks James! I have tried cluster.addClientFixture(), due to this method doesn't copy all the client properties from ClusterFixture, so after create a new ClientFixture, it still has problem to connect with cluster. And I also tried to modify updateClient method signature to return a client, however due to the cluster started by startCluster() is hold by a static variable, and updateClient should be a non-static method to refactor to ClientFixture, and seems a non-static method return a value to static variable made some error here, and I am still figuring it out. I think the feasible way should be start ClientFixture and ClusterFixture separately in each case, so that those two variables are both non-static. The drawback the is the code change should be a little too much, but I can submit it commit to see whether it works

Reply:
Hi James, I pushed a new commit which refactor updateClient from ClusterTest to ClientFixture, and I can pass the client into updateClient so that we don't need to client.close for each updateClient, how method signature needs add  another variable specifically for close client.